typescript: Migrate pages dir to typescript.

vsvipul opened the pull request:
**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

vsvipul committed:
gitignore: Add pages/ dir to gitignore.

vsvipul committed:
typescript: Migrate add-certificate.js to typescript.

vsvipul committed:
typescript: Migrate badge-settings.js to typescript.

vsvipul committed:
typescript: Migrate base-section.js to typescript.

vsvipul committed:
typescript: Migrate connected-org-section.js to typescript.

vsvipul committed:
typescript: Migrate find-accounts.js to typescript.

vsvipul committed:
typescript: Migrate general-section.js to typescript.

vsvipul committed:
typescript: Migrate nav.js to typescript.

vsvipul committed:
typescript: Migrate network-section.js to typescript.

vsvipul committed:
typescript: Migrate new-server-form.js to typescript.

vsvipul committed:
typescript: Migrate preference.js to typescript.

vsvipul committed:
typescript: Migrate server-info-form.js to typescript.

vsvipul committed:
typescript: Migrate servers-section.js to typescript.

vsvipul committed:
typescript: Migrate shortcuts-section.js to typescript.

vsvipul committed:
typescript: Migrate network.js to typescript.

priyank-p commented on app/renderer/js/pages/preference/add-certificate.ts:
```diff
@@ -1,20 +1,28 @@
+	_certFile: string;
+	$addCertificate: Element | null;
+	addCertificateButton: Element | null;
+	serverUrl: Element | null;
```
Make this `HTMLInputElement` since we know it's a input element. That way we don't need to type cast this to `HTMLInputElement`.

priyank-p commented on app/renderer/js/pages/preference/add-certificate.ts:
```diff
@@ -52,35 +60,35 @@ class AddCertificate extends BaseComponent {
 			filters: [{ name: 'crt, pem', extensions: ['crt', 'pem'] }]
 		};
-		dialog.showOpenDialog(showDialogOptions, selectedFile => {
+		dialog.showOpenDialog(showDialogOptions as OpenDialogOptions, selectedFile => {
```
It's better to do `const showDialogOptions: OpenDialogOptions` above where the variable is declared. Also typescript should automatically be type-checking for `showDialogOptions` anyway so we probably don't need to do this?

vsvipul replied:
tsc was giving an error on this. So, added this then. I'll try ```const showDialogOptions: OpenDialogOptions``` .

vsvipul replied:
```
'defaultId' does not exist in type 'OpenDialogOptions'.
```
This was the problem here. So doing it as i did it was necessary.

priyank-p replied:
TypeScript here is correct `defaultId` is only avalible for `showMessageBox` not`showOpenDialog` so we should remove it. Also can you submit this patch to master?

vsvipul replied:
OK, great.! will do.

priyank-p commented on app/renderer/js/pages/preference/add-certificate.ts:
```diff
@@ -52,35 +60,35 @@ class AddCertificate extends BaseComponent {
 
 		this.serverUrl.addEventListener('keypress', event => {
-			const EnterkeyCode = event.keyCode;
+			const EnterkeyCode = (event as KeyboardEvent).keyCode;
```
You should do `(event: KeyboardEvent)` above; `as` is generally used for type assertion.

vsvipul replied:
Wasn't required. Removed it.

priyank-p commented on app/renderer/js/pages/preference/badge-settings.ts:
```diff
@@ -1,10 +1,10 @@
+import { app } from 'electron';
 
-const ConfigUtil = require(__dirname + '/../../utils/config-util.js');
+import electron = require('electron');
```
Minor suggestion here, you can also do `import electron, { app } from 'electron';` I am pretty sure.

vsvipul replied:
Gave the powerMonitor error again after using this.

priyank-p replied:
Oh, ok!

priyank-p commented on app/renderer/js/pages/preference/connected-org-section.ts:
```diff
@@ -1,18 +1,27 @@
+	props: any;
+	$serverInfoContainer: Element | null;
+	$existingServers: Element | null;
+	$newOrgButton: Element | null;
```
Since we know this a button make this a `HTMLButtonElement` so later if we needed to do say `.click` somebody doesn't need to fiddle with this.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.ts:
```diff
@@ -24,15 +29,15 @@ class FindAccounts extends BaseComponent {
+		this.props.$root.append(this.$findAccounts);
 		this.$findAccountsButton = this.$findAccounts.querySelector('#find-accounts-button');
-		this.$serverUrlField = this.$findAccounts.querySelectorAll('input.setting-input-value')[0];
+		this.$serverUrlField = this.$findAccounts.querySelectorAll('input.setting-input-value')[0] as HTMLInputElement;
```
Do we really need to do `as HTMLInputElement` here? I feel like this should be unnessacary since we declared above that this variable is in fact type of `HTMLInputElement`.

vsvipul replied:
Then, we get this-
```
Type 'Element' is missing the following properties from type 'HTMLInputElement': accept, align, alt, autocomplete, and 164 more.ts(2740)
```

priyank-p replied:
OK, so we do need this.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.ts:
```diff
@@ -42,14 +47,14 @@ class FindAccounts extends BaseComponent {
 		this.$serverUrlField.addEventListener('click', () => {
 			if (this.$serverUrlField.value === 'zulipchat.com') {
-				this.$serverUrlField.setSelectionRange(0, 0);
+				(this.$serverUrlField.setSelectionRange(0, 0));
```
This is syntax error, since there is a open paren here?

priyank-p commented on app/renderer/js/pages/preference/general-section.ts:
```diff
@@ -354,30 +359,30 @@ class GeneralSection extends BaseSection {
 		};
 
-		dialog.showOpenDialog(showDialogOptions, selectedFile => {
+		dialog.showOpenDialog(showDialogOptions as OpenDialogOptions, selectedFile => {
```
See my comment above for the similar thing in previous commit.

vsvipul replied:
defaultId is not a property in OpenDialogOptions. Should i remove it?

priyank-p replied:
Yes, I mentioned this above for similar thing.

priyank-p commented on app/renderer/js/pages/preference/general-section.ts:
```diff
@@ -389,51 +394,51 @@ class GeneralSection extends BaseSection {
-			const cssPATH = document.getElementById('remove-custom-css');
-			cssPATH.style.display = 'none';
+			const cssPATH = document.querySelector('#remove-custom-css');
+			(cssPATH as HTMLElement).style.display = 'none';
```
Do `const cssPATH: HTMLElement = ...` in the previous line.

priyank-p commented on app/renderer/js/pages/preference/general-section.ts:
```diff
@@ -389,51 +394,51 @@ class GeneralSection extends BaseSection {
+		const removeCSSButton = document.querySelector('#css-delete-action');
 		removeCSSButton.addEventListener('click', () => {
-			ConfigUtil.setConfigItem('customCSS');
+			ConfigUtil.setConfigItem('customCSS', undefined);
```
Let's make this `""`, a blank string, since we later pull out this value an inject it into the style tag anyways. cc @akashnimare

priyank-p commented on app/renderer/js/pages/preference/general-section.ts:
```diff
@@ -389,51 +394,51 @@ class GeneralSection extends BaseSection {
 		};
 
-		dialog.showOpenDialog(showDialogOptions, selectedFolder => {
+		dialog.showOpenDialog(showDialogOptions as OpenDialogOptions, selectedFolder => {
```
Ditto.

priyank-p commented on app/renderer/js/pages/preference/general-section.ts:
```diff
@@ -389,51 +394,51 @@ class GeneralSection extends BaseSection {
 				ConfigUtil.setConfigItem('downloadsPath', selectedFolder[0]);
 				const downloadFolderPath = document.querySelector('.download-folder-path');
-				downloadFolderPath.innerText = selectedFolder[0];
+				(downloadFolderPath as HTMLElement).innerText = selectedFolder[0];
```
Ditto.

priyank-p commented on app/renderer/js/pages/preference/new-server-form.ts:
```diff
@@ -36,21 +42,20 @@ class NewServerForm extends BaseComponent {
-
-		this.$newServerUrl = this.$newServerForm.querySelectorAll('input.setting-input-value')[0];
+		this.props.$root.append(this.$newServerForm);
+		this.$newServerUrl = this.$newServerForm.querySelectorAll('input.setting-input-value')[0] as HTMLInputElement;
```
See my comment about this in previous commit.

vsvipul replied:
This is required because querySelector returns Element type.

priyank-p replied:
OK, sounds good!

priyank-p commented on app/renderer/js/pages/preference/preference.ts:
```diff
@@ -1,24 +1,28 @@
+	$settingsContainer: Element;
+	nav: typeof Nav;
+	section: typeof ServersSection | typeof GeneralSection |
+		typeof NetworkSection | typeof ConnectedOrgSection | typeof ShortcutsSection;
```
Let extract all of this out to a variable, i.e. `type Section = typeof ...` and use it there.

priyank-p commented on app/renderer/js/pages/preference/preference.ts:
```diff
@@ -77,32 +81,32 @@ class PreferenceView extends BaseComponent {
 		const input = document.querySelector(inputSelector);
 		if (input) {
-			input.checked = state;
+			(input as HTMLInputElement).checked = state;
```
Ditto.

priyank-p commented on app/renderer/js/pages/preference/preference.ts:
```diff
@@ -77,32 +81,32 @@ class PreferenceView extends BaseComponent {
-	registerIpcs() {
-		ipcRenderer.on('switch-settings-nav', (event, navItem) => {
+	registerIpcs(): void {
+		ipcRenderer.on('switch-settings-nav', (_event: Event, navItem: string) => {
```
I have been meaning to turn off this rule that make use `_` before unused parameter. Can you look into this?

vsvipul replied:
We have done this while migrating other files as well. Will have to change it after everything is done later, if we want to do this.

priyank-p replied:
Actually, we can leave the already done as is with `_`, if we turn this lint rule off we just don't have to keep doing this.

vsvipul replied:
Ok, we can then turn this off in a later commit!

priyank-p commented on app/renderer/js/pages/preference/shortcuts-section.ts:
```diff
@@ -0,0 +1,341 @@
+		this.props = props;
+	}
+
+	templateMac(): string {
```
Can you add a todo here briefly explaining that we need to de-duplicate the `templateMac()` and `templateWinLin()` here. Both of them have same template literal but and the only thing diffrent about them is the `userOSKey` which we could pass as argument.

vsvipul replied:
No, the template literal is not same. I just compared them. One is 157 lines, while other is 141. Still, added a comment.

priyank-p replied:
Hmm cool! Either way I suspect it majority of it is the same. Thanks for adding the comment.

priyank-p reviewed (commented):
@vsvipul posted a bunch of comments. Can you address those?

vsvipul commented:
@priyank-p I have made some changes. Can you please do a review again ?

vsvipul committed:
tsconfig: Turn off noUnusedParamaters rule.

We want to turn off this rule because this requires us to add _
in front of parameter that are unused which we have lot of, for example,
the ipc listeners.

priyank-p commented on app/renderer/js/pages/preference/connected-org-section.ts:
```diff
@@ -57,26 +66,24 @@ class ConnectedOrgSection extends BaseSection {
 		this.$newOrgButton.addEventListener('click', () => {
-			// We don't need to import this since it's already imported in other files
-			// eslint-disable-next-line no-undef
 			ipcRenderer.send('forward-message', 'open-org-tab');
```
I wonder how this didn't throw an error before!

priyank-p commented:
@vsvipul I extend the todo comment a bit for the `shortcut-section.ts` and merged this along with other PRs. We just finished the migration of all the files to typescript.

Next thing, we should work on the gulp thing we talked about, then work on making `npm start` workflow more robuts and lastly the moving the compiled js files to `src`. Once again @vsvipul thanks for working on this!
